Svc/FpySequencer/FpySequencerDirectives.cpp

```diff
@@ -28,8 +28,128 @@ void FpySequencer::sendSignal(Signal signal) {
     }
 }
 
-I64& FpySequencer::reg(U8 idx) {
-    return this->m_runtime.regs[idx];
+template <typename T>
+T FpySequencer::pop() {
+    static_assert(sizeof(T) == 8 || sizeof(T) == 4 || sizeof(T) == 2 || sizeof(T) == 1, "size must be 1, 2, 4, 8");
+    FW_ASSERT(this->m_runtime.stackSize >= sizeof(T), static_cast<FwAssertArgType>(this->m_runtime.stackSize), static_cast<FwAssertArgType>(sizeof(T)));
+    // first make a byte array which can definitely store our val
+    U8 valBytes[8] = {0};
+    // now move top of stack into byte array and shrink stack
+    memcpy(valBytes, this->top() - sizeof(T), sizeof(T));
+    this->m_runtime.stackSize -= sizeof(T);
+
+    // now do appropriate byteswap on byte array
+    if (sizeof(T) == 8) {
+        return static_cast<T>(
+               (static_cast<T>(valBytes[7]) << 0)  |
+               (static_cast<T>(valBytes[6]) << 8)  |
+               (static_cast<T>(valBytes[5]) << 16) |
+               (static_cast<T>(valBytes[4]) << 24) |
+               (static_cast<T>(valBytes[3]) << 32) |
+               (static_cast<T>(valBytes[2]) << 40) |
+               (static_cast<T>(valBytes[1]) << 48) |
+               (static_cast<T>(valBytes[0]) << 56));
+    } else if (sizeof(T) == 4) {
+        return static_cast<T>(
+               (static_cast<T>(valBytes[3]) << 0) |
+               (static_cast<T>(valBytes[2]) << 8) |
+               (static_cast<T>(valBytes[1]) << 16) |
+               (static_cast<T>(valBytes[0]) << 24));
+    } else if (sizeof(T) == 2) {
+        return static_cast<T>(
+               (static_cast<T>(valBytes[1]) << 0) |
+               (static_cast<T>(valBytes[0]) << 8));
+    } else {
+        return static_cast<T>(valBytes[0]);
+    }
+}
+
+template U8 FpySequencer::pop();
+template U16 FpySequencer::pop();
+template U32 FpySequencer::pop();
+template U64 FpySequencer::pop();
+template I8 FpySequencer::pop();
+template I16 FpySequencer::pop();
+template I32 FpySequencer::pop();
+template I64 FpySequencer::pop();
+
+template <> F32 FpySequencer::pop<F32>() {
+    U32 endianness = this->pop<U32>();
+    F32 val;
+    memcpy(&val, &endianness, sizeof(val));
+    return val;
+}
+
+template <> F64 FpySequencer::pop<F64>() {
+    U64 endianness = this->pop<U64>();
+    F64 val;
+    memcpy(&val, &endianness, sizeof(val));
+    return val;
+}
+
+template <typename T>
+void FpySequencer::push(T val) {
```

Comment:
This code looks like it may be doing right shifts of signed values. The result of that operation is not defined in C++ and should be avoided. It's better to do the twos-complement arithmetic explicitly, and only shift unsigned values.

Reply:
You should be able to fix this by casting between signed and unsigned values, and shifting unsigned values. I don't think any complicated arithmetic is required. You can use `std::make_unsigned` to get the unsigned type with the same bit width as `T`.